packages/react-async/src/useAsync.tsx

```diff
@@ -135,52 +181,97 @@ const useAsync = (arg1, arg2) => {
     if (counter.current) cancel()
     if (promise || promiseFn) load()
   }, [promise, promiseFn, watch])
-  useEffect(() => () => (isMounted.current = false), [])
+  useEffect(
+    () => () => {
+      isMounted.current = false
+    },
+    []
+  )
   useEffect(() => () => cancel(), [])
   /* eslint-enable react-hooks/exhaustive-deps */
 
   useDebugValue(state, ({ status }) => `[${counter.current}] ${status}`)
 
   return useMemo(
-    () => ({
-      ...state,
-      run,
-      reload,
-      cancel,
-      setData,
-      setError,
-    }),
+    () =>
+      ({
+        ...state,
+        run,
+        reload,
+        cancel,
+        setData,
+        setError,
+      } as AsyncState<T>),
     [state, run, reload, cancel, setData, setError]
   )
 }
 
-const parseResponse = (accept, json) => res => {
+const parseResponse = (accept: undefined | string, json: undefined | boolean) => (
+  res: Response
+) => {
   if (!res.ok) return Promise.reject(res)
   if (typeof json === "boolean") return json ? res.json() : res
   return accept === "application/json" ? res.json() : res
 }
 
-const useAsyncFetch = (input, init, { defer, json, ...options } = {}) => {
-  const method = input.method || (init && init.method)
-  const headers = input.headers || (init && init.headers) || {}
-  const accept = headers["Accept"] || headers["accept"] || (headers.get && headers.get("accept"))
-  const doFetch = (input, init) => globalScope.fetch(input, init).then(parseResponse(accept, json))
+interface FetchRun<T> extends Omit<AbstractState<T>, "run"> {
+  run(overrideInit: (init: RequestInit) => RequestInit): void
+  run(overrideInit: Partial<RequestInit>): void
+  run(ignoredEvent: React.SyntheticEvent): void
+  run(ignoredEvent: Event): void
+  run(): void
+}
+
+/**
+ *
+ * @param {RequestInfo} input
+ * @param {RequestInit} init
+ * @param {FetchOptions} options
+ * @returns {AsyncState<T, FetchRun<T>>}
+ */
+const useAsyncFetch = <T extends {}>(
+  input: RequestInfo,
+  init: RequestInit,
+  { defer, json, ...options }: FetchOptions<T> = {}
+): AsyncState<T, FetchRun<T>> => {
+  const method = (input as Request).method || (init && init.method)
+  const headers: Headers & Record<string, any> =
+    (input as Request).headers || (init && init.headers) || {}
+  const accept: string | undefined =
+    headers["Accept"] || headers["accept"] || (headers.get && headers.get("accept"))
+  const doFetch = (input: RequestInfo, init: RequestInit) =>
+    globalScope.fetch(input, init).then(parseResponse(accept, json))
   const isDefer =
-    typeof defer === "boolean" ? defer : ["POST", "PUT", "PATCH", "DELETE"].indexOf(method) !== -1
+    typeof defer === "boolean" ? defer : ["POST", "PUT", "PATCH", "DELETE"].indexOf(method!) !== -1
```

Comment:
@phryneas IMO enums can be used over here as these strings(POST, PUT, PATCH, DELETE) are not going to change. what I mean is doing something like this 
```js
const {POST, PUT, PATCH, DELETE} = Method //enum 
[POST, PUT, PATCH, DELETE].indexOf(method!) !== -1

```